### DISCUSSION TITLE
How to create a global/catch-all error

### DISCUSSION THREAD

**Discussion by jojohanhannesnes** [tokio-rs/axum#2266] [Q&A] [answered]
Hi,
I'm currently learning axum and having a hard time trying to create a global/catch-all error on my routes.
I want to achieve something similiar in expressjs where i can just put another routes/layer at the end of the router that can catch every issues on the app. I have found something similiar like `fallback`, just wanted some routes that's get triggered when there's an error (even before the request gets to the handler).

The reason i wanted to implement this is because i found an error when writing the route, if i put the layer and routes in the wrong order, when hitting the api i'll be getting some error with like ``Missing request extension: Extension of type `entity::user::Model` was not found. Perhaps you forgot to add it? See `axum::Extension` ``. I wanted to convert that (or any other kind of error) message into an custom error response that i already have 

Here's my current code, for context

```rust
// main.rs
async fn server() {
    let connection_string = (utils::constants::DATABASE_URL).clone();
    let db: DatabaseConnection = Database::connect(connection_string).await.unwrap();
    init_cron();
    tracing_subscriber::fmt().with_target(false).init();

    let app: Router = Router::new()
        .route("/", get(|| async {}))
        .merge(routes::user::routes())
        .merge(routes::expenses::routes())
        .route_layer(middleware::from_fn(utils::guards::guard))
        .merge(routes::expenses::routes())
        .merge(routes::auth::routes())
        .layer(Extension(db))
        .layer(
            TraceLayer::new_for_http()
                .make_span_with(trace::DefaultMakeSpan::new().level(Level::INFO))
                .on_response(trace::DefaultOnResponse::new().level(Level::INFO)),
        )
        .fallback(handler_404);
        // wanted to add the catch-all error handler here
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    tracing::info!("listening on {}", addr);
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap()
}

async fn handler_404() -> impl IntoResponse {
    (
        StatusCode::NOT_FOUND,
        "The requested resource was not found",
    )
}
```

```rust
// error.rs
use axum::{
    http::{header, StatusCode, Method, Uri},
    response::IntoResponse,
    Json, BoxError,
};
use serde_json::json;
use tower::timeout::error::Elapsed;

pub struct APIError {
    pub message: String,
    pub status_code: StatusCode,
    pub error_code: Option<i8>,
}

impl IntoResponse for APIError {
    fn into_response(self) -> axum::response::Response {
        let status_code = self.status_code;
        (status_code, 
            [(header::CONTENT_TYPE, "application/json")],
            Json(json!({"StatusCode": self.status_code.as_u16(), "ErrorCode": self.error_code, "Message": self.message}))
        ).into_response()
    }
}
```
this is the [example](https://expressjs.com/en/guide/error-handling.html) i'm trying to duplicate
```javascript
function errorHandler (err, req, res, next) {
  res.status(500)
  res.render('error', { error: err })
}
```

would appreciate any help, thanks!

**Comment by engineerhead** [comment]
I am in the same boat as yours. Looking to find a way to have a global middleware.

**Reply by antmelnyk** [reply]
Same. Was looking into a simple way to handle all deserialize errors from extractors in one place.

**Comment by davidpdrsn** [comment, marked as answer]
axum currently doesn’t have a way catch any possible error.

See https://github.com/tokio-rs/axum/tree/main/examples/customize-extractor-error for how to customize the response of extractors.

Generally you shouldn’t need to customize the response from Extension because that error should _never_ reach the end user. That is always a programmer error that should be caught by your automated tests.

Id also recommend using https://docs.rs/axum/0.6.20/axum/struct.Router.html#method.with_state instead of Extension.

**Reply by jojohanhannesnes** [reply]
Hi @davidpdrsn , thanks for the response, i assume the recommended link is also related to this #1830? i'd take a look at both links you mentioned. Once again thanks!
